File: php/src/PlaybackBuilder.php
<?php

declare(strict_types=1);

namespace Servirtium;

use FFI;
use FFI\CData;

/** Configures and starts a playback VCR server. */
final class PlaybackBuilder extends VcrBuilderBase
{
    /** @var list<array{0: VcrField, 1: string, 2: string}> */
    private array $unredactions = [];

    private bool $strictHeadersOn = false;

    private bool $matchJsonBodyOn = false;

    private bool $matchMultipleOn = false;

    /** @var list<string> */
    private array $matchHeaders = [];

    public function __construct(string $tapePath)
    {
        parent::__construct($tapePath);
    }

    /**
     * Compare the SUT's request headers against the recorded block on every
     * interaction (Servirtium step 10), surfacing mismatches via
     * {@see VcrServer::lastError()}.
     */
    public function strictHeaders(bool $on = true): static
    {
        $this->strictHeadersOn = $on;

        return $this;
    }

    /**
     * Opt in to matching request bodies by semantic JSON equality (key order /
     * whitespace ignored) instead of byte-for-byte. Non-JSON bodies fall back
     * to byte-exact.
     */
    public function matchJsonBody(bool $on = true): static
    {
        $this->matchJsonBodyOn = $on;

        return $this;
    }

    /**
     * Opt in to reusable, order-independent playback: matches any recorded
     * interaction (not just the next in sequence) and doesn't consume it — for
     * polling/retries or non-deterministic request order.
     */
    public function matchMultiple(bool $on = true): static
    {
        $this->matchMultipleOn = $on;

        return $this;
    }

    /**
     * Match playback on this specific request header's value (ignoring the rest
     * of the recorded header block); repeatable.
     */
    public function matchHeader(string $name): static
    {
        $this->matchHeaders[] = $name;

        return $this;
    }

    /**
     * Replace a redacted placeholder in the recorded expectation with the
     * real value the live SUT sends, so a committed (scrubbed) tape still
     * matches.
     */
    public function unredact(VcrField $field, string $pattern, string $replacement): static
    {
        $this->unredactions[] = [$field, $pattern, $replacement];

        return $this;
    }

    protected function applyConfig(CData $handle): void
    {
        parent::applyConfig($handle);
        $lib = Native::lib();
        if ($this->strictHeadersOn) {
            $lib->aether_vcr_embed_set_strict_headers($handle, 1);
        }
        if ($this->matchJsonBodyOn) {
            $lib->aether_vcr_embed_set_match_json_body($handle, 1);
        }
        if ($this->matchMultipleOn) {
            $lib->aether_vcr_embed_set_match_multiple($handle, 1);
        }
        foreach ($this->matchHeaders as $name) {
            $lib->aether_vcr_embed_match_header($handle, $name);
        }
        foreach ($this->unredactions as [$field, $pattern, $replacement]) {
            self::check($lib->aether_vcr_embed_unredact($handle, $field->value, $pattern, $replacement), 'unredact');
        }
    }

    public function start(): VcrServer
    {
        Native::configure($this->nativeLibValue);
        $lib = Native::lib();
        $handle = $lib->aether_vcr_embed_open_playback(
            $this->labelValue,
            $this->tapePath,
            $this->hostValue,
            $this->portValue,
        );
        if ($handle === null || FFI::isNull($handle)) {
            throw new VcrException("vcr playback failed to start for tape '{$this->tapePath}'");
        }
        $this->applyConfig($handle);
        if ($lib->aether_vcr_embed_start($handle) < 0) {
            $err = self::drainStartError($handle);
            $lib->aether_vcr_embed_stop($handle);
            throw new VcrException("vcr playback failed to begin serving for tape '{$this->tapePath}': {$err}");
        }

        return new VcrServer($handle, $this->hostValue, $this->tapePath, recordMode: false, failIfChanged: false);
    }
}

File: java/src/main/java/com/paulhammant/servirtium/vcr/PlaybackBuilder.java
package com.paulhammant.servirtium.vcr;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.util.ArrayList;
import java.util.List;

/** Configures and starts a playback VCR server. */
public final class PlaybackBuilder extends VcrBuilderBase<PlaybackBuilder> {

    private record Unredaction(Field field, String pattern, String replacement) {
    }

    private final List<Unredaction> unredactions = new ArrayList<>();
    private boolean strictHeaders;
    private boolean matchJsonBody;

    PlaybackBuilder(String tapePath) {
        super(tapePath);
    }

    @Override
    PlaybackBuilder self() {
        return this;
    }

    /**
     * Compare the SUT's request headers against the recorded block on every
     * interaction (Servirtium step 10), surfacing mismatches via
     * {@link VcrServer#lastError()}.
     */
    public PlaybackBuilder strictHeaders() {
        return strictHeaders(true);
    }

    public PlaybackBuilder strictHeaders(boolean on) {
        this.strictHeaders = on;
        return this;
    }

    /**
     * Opt in to matching request bodies by <em>semantic</em> JSON equality (key
     * order and insignificant whitespace ignored) instead of byte-for-byte.
     * Non-JSON bodies fall back to the byte-exact comparison, so this never
     * loosens matching for non-JSON payloads.
     */
    public PlaybackBuilder matchJsonBody() {
        return matchJsonBody(true);
    }

    public PlaybackBuilder matchJsonBody(boolean on) {
        this.matchJsonBody = on;
        return this;
    }

    /**
     * Replace a redacted placeholder in the recorded expectation with the real
     * value the live SUT sends, so a committed (scrubbed) tape still matches.
     */
    public PlaybackBuilder unredact(Field field, String pattern, String replacement) {
        unredactions.add(new Unredaction(field, pattern, replacement));
        return this;
    }

    @Override
    void applyConfig(Arena arena, MemorySegment handle) {
        super.applyConfig(arena, handle);
        try {
            if (strictHeaders) {
                NativeMethods.SET_STRICT_HEADERS.invokeExact(handle, 1);
            }
            if (matchJsonBody) {
                NativeMethods.SET_MATCH_JSON_BODY.invokeExact(handle, 1);
            }
            for (Unredaction u : unredactions) {
                MemorySegment r = (MemorySegment) NativeMethods.UNREDACT.invokeExact(
                        handle,
                        u.field().code(),
                        NativeMethods.cString(arena, u.pattern()),
                        NativeMethods.cString(arena, u.replacement()));
                check(r, "unredact");
            }
        } catch (Throwable t) {
            throw rethrow("applyConfig", t);
        }
    }

    public VcrServer start() {
        NativeLoader.configure(nativeLib);
        MemorySegment handle;
        try (Arena arena = Arena.ofConfined()) {
            handle = (MemorySegment) NativeMethods.OPEN_PLAYBACK.invokeExact(
                    NativeMethods.cString(arena, label),
                    NativeMethods.cString(arena, tapePath),
                    NativeMethods.cString(arena, host),
                    port);
            if (handle.address() == 0) {
                throw new VcrException("vcr playback failed to start for tape '" + tapePath + "'");
            }
            applyConfig(arena, handle);
            int rc = (int) NativeMethods.START.invokeExact(handle);
            if (rc < 0) {
                String detail = drainStartError(handle);
                NativeMethods.STOP.invokeExact(handle);
                throw new VcrException(
                        "vcr playback failed to begin serving for tape '" + tapePath + "': " + detail);
            }
        } catch (Throwable t) {
            throw rethrow("playback start", t);
        }
        return new VcrServer(handle, host, tapePath, false, false);
    }
}
